go-runtime/compile/build.go

```diff
@@ -48,45 +48,155 @@ type ExternalModuleContext struct {
 	Replacements []*modfile.Replace
 }
 
-type goVerb struct {
-	Name        string
-	Package     string
-	MustImport  string
-	HasRequest  bool
-	HasResponse bool
-}
-
 type mainModuleContext struct {
 	GoVersion          string
 	FTLVersion         string
 	Name               string
-	ProjectName        string
 	SharedModulesPaths []string
 	Verbs              []goVerb
 	Replacements       []*modfile.Replace
-	SumTypes           []goSumType
-	LocalSumTypes      []goSumType
-	ExternalTypes      []goExternalType
-	LocalExternalTypes []goExternalType
+	MainCtx            mainFileContext
+	TypesCtx           typesFileContext
+}
+
+func (c mainModuleContext) withImports() mainModuleContext {
+	c.MainCtx.Imports = c.generateMainImports()
+	c.TypesCtx.Imports = c.generateTypesImports()
+	return c
+}
+
+func (c mainModuleContext) generateMainImports() []string {
+	imports := sets.NewSet[string]()
+	imports.Add(`"context"`)
+	imports.Add(`"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"`)
+	imports.Add(`"github.com/TBD54566975/ftl/common/plugin"`)
+	imports.Add(`"github.com/TBD54566975/ftl/go-runtime/server"`)
+	if len(c.MainCtx.SumTypes) > 0 || len(c.MainCtx.ExternalTypes) > 0 {
+		imports.Add(`"github.com/TBD54566975/ftl/go-runtime/ftl/reflection"`)
+	}
+
+	for _, v := range c.Verbs {
+		imports.Add(v.importStatement())
+	}
+	for _, st := range c.MainCtx.SumTypes {
+		imports.Add(st.importStatement())
+
+		for _, v := range st.Variants {
+			if i := strings.LastIndex(v.Type, "."); i != -1 {
+				lessTypeName := strings.TrimSuffix(v.Type, v.Type[i:])
+				imports.Add(strconv.Quote(lessTypeName))
+			}
+		}
+	}
+	for _, e := range c.MainCtx.ExternalTypes {
+		imports.Add(e.importStatement())
+	}
+	out := imports.ToSlice()
+	slices.Sort(out)
+	return out
+}
+
+func (c mainModuleContext) generateTypesImports() []string {
+	imports := sets.NewSet[string]()
+	if len(c.TypesCtx.SumTypes) > 0 || len(c.TypesCtx.ExternalTypes) > 0 {
+		imports.Add(`"github.com/TBD54566975/ftl/go-runtime/ftl/reflection"`)
+	}
+	for _, st := range c.TypesCtx.SumTypes {
+		// subpackage
```

Comment:
nit: would you mind making these two comments something like `// if import path is more than 2 dirs deep, it's a subpackage`? It feels obvious now that I've grokked it, but it took me a minute to figure out initially. Already super helpful having the existing comment there anyway so thanks :D